graviton/sim.py

```diff
@@ -0,0 +1,116 @@
+from dataclasses import dataclass
+from typing import Any, Dict, Iterable, List, Optional, Sequence, TypeVar, Union, cast
+
+from algosdk.v2client.algod import AlgodClient
+
+from graviton.abi_strategy import ABICallStrategy
+from graviton.blackbox import DryRunExecutor, DryRunTransactionParams as TxParams
+from graviton.inspector import DryRunProperty as DRProp, DryRunInspector
+from graviton.invariant import Invariant
+from graviton.models import ExecutionMode, PyTypes
+
+# TODO: this will encompass strategies, composed of
+# hypothesis strategies as well as home grown ABIStrategy sub-types
+InputStrategy = Union[Iterable[Sequence[PyTypes]], ABICallStrategy]
+
+
+@dataclass(frozen=True)
+class SimulationResults:
+    succeeded: bool
+    simulate_inspectors: List[DryRunInspector]
+    identities_inspectors: Optional[List[DryRunInspector]] = None
+
+
+class Simulation:
+    """
+    Simulation ~ (Teal logic for execution) + (predicates that must be satisfied)
+
+    TODO: Slated to become the hypothesis plugin receptor.
+    TODO: Add some real comments (cf. Issue #51)
+    """
+
+    def __init__(
+        self,
+        algod: AlgodClient,
+        mode: ExecutionMode,
+        simulate_teal: str,
+        predicates: Dict[DRProp, Any],
+        *,
+        abi_method_signature: Optional[str] = None,
+        omit_method_selector: bool = False,
+        validation: bool = True,
+        identities_teal: Optional[str] = None,
+    ):
+        self.simulate_dre: DryRunExecutor = DryRunExecutor(
+            algod,
+            mode,
+            simulate_teal,
+            abi_method_signature=abi_method_signature,
+            omit_method_selector=omit_method_selector,
+            validation=validation,
+        )
+        self.identities_dre: Optional[DryRunExecutor] = None
+        if identities_teal:
```

Comment:
I might remember or understand it wrong, is this for cross version check (i.e.., ver6 vs v8)?

Reply:
👍